soroban-env-host/src/crypto/bn254.rs

```diff
@@ -36,30 +37,92 @@ impl Host {
         self.err(ScErrorType::Crypto, ScErrorCode::InvalidInput, msg, &[])
     }
 
-    // This is the internal routine performing deserialization on various
-    // element types, which can be conceptually decomposed into units of Fp
-    // (the base field element), and will be charged accordingly.
-    // Validation of the deserialized entity must be performed outside of this
-    // function, to keep budget charging isolated.
-    pub(crate) fn bn254_deserialize_uncompressed_no_validate<
+    // Arkworks and ethereum differ in terms of how flags are handled.
+    //
+    // - In arkworks, the two free bits are reserved for flags: the
+    //   most-significant-bit is the Y-sign flag, and the 2nd
+    //   most-significant-bit is the infinity flag
+    // - In ethereum, the two free bits are always unset.
+    //
+    // Since encoding/decoding is in uncompressed mode (Y is included), there is
+    // no disambiguity. But we need to make sure to stick to the our
+    // specification (which is modeled after Ethereum). That means during
+    // decoding, we make sure the incoming bytes have the two "flag bits" remain
+    // unset.
+    fn bn254_validate_flags_and_check_infinity<const EXPECTED_SIZE: usize>(
+        &self,
+        bytes: &[u8],
+        tag: &str,
+    ) -> Result<bool, HostError> {
+        if bytes.len() != EXPECTED_SIZE {
+            return Err(self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                format!("bn254 {}: invalid input length to deserialize", tag).as_str(),
+                &[
+                    Val::from_u32(bytes.len() as u32).into(),
+                    Val::from_u32(EXPECTED_SIZE as u32).into(),
+                ],
+            ));
+        }
+
+        // The incoming bytes is in big-endian, the sign bits are contained in
+        // the first byte (for G1Affine that's the MSB of the y-coordinate (Fp),
+        // for G2Affine that's the MSB of the c1 (Fp) part of the Y-coordinate
+        // (Fp2)). The highest bit (0x80) is the y-sign flag, the 2nd highest
+        // bit (0x40) is the infinity flag. We ensure both are unset.
+        let flags = 0b1100_0000 & bytes[0];
+        if flags != 0b0000_0000 {
+            return Err(self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                format!("bn254 {} deserialize: the two flag bits must be unset", tag).as_str(),
+                &[],
+            ));
+        }
+        let inf = bytes.iter().all(|b| *b == 0u8);
```

Comment:
Can we add some comment on infinity representation? For some reason I thought it was all 0's...